documentation/src/main/asciidoc/user_guide/integrations_hibernate_cache.adoc

```diff
@@ -0,0 +1,506 @@
+
+.Enable second-level cache
+[cols="1,10"]
+|===
+| JPA       | `<property name="hibernate.cache.use_second_level_cache" value="true"/>`
+| Spring    | `spring.jpa.properties.hibernate.cache.use_second_level_cache=true`     
+|===
+
+To select which entities/collections to cache, first annotate them with `javax.persistence.Cacheable`.
+Then make sure shared cache mode is set to `ENABLE_SELECTIVE`:
+
+.Enable selective shared cached mode
+[cols="1,10"]
+|===
+| JPA       | `<shared-cache-mode>ENABLE_SELECTIVE</shared-cache-mode>`
+| Spring    | `spring.jpa.properties.javax.persistence.sharedCache.mode=ENABLE_SELECTIVE` 
+|===
+
+NOTE: This is the most common way of selecting which entities/collections to cache.
+However, there are alternative ways to which are explained in the
+link:https://docs.jboss.org/hibernate/orm/5.2/userguide/html_single/Hibernate_User_Guide.html#caching-mappings[Hibernate documentation].
+
+Optionally, queries can also be cached but for that query cache needs to be enabled:
+
+.Enable query cache
+[cols="1,10"]
+|===
+| JPA       | `<property name="hibernate.cache.use_query_cache" value="true"/>`
+| Spring    | `spring.jpa.properties.hibernate.cache.use_query_cache=true`     
+|===
+
+NOTE: As well as enabling query cache, forcing a query to be cached requires the query to be made cacheable.
+For example, for JPA queries: `query.setHint("org.hibernate.cacheable", Boolean.TRUE)`.
+
+The best way to find out whether second level cache is working or not is to inspect the statistics.
+By inspecting the statistics you can verify if the cache is being hit, if any new data is stored in cache...etc.
+Statistics are disabled by default, so it is recommended that you enable statistics:
+
+.Enable statistics
+[cols="1,10"]
+|===
+| JPA       | `<property name="hibernate.generate_statistics" value="true" />`
+| Spring    | `spring.jpa.properties.hibernate.generate_statistics=true`      
+|===
+
+
+==== Deployment Scenarios
+
+How to configure Infinispan to be the second level cache provider varies slightly depending on the deployment scenario:
+
+===== Single-Node Standalone Hibernate Application
+
+In standalone library mode, a JPA/Hibernate application runs inside a Java SE application or inside containers that don’t offer Infinispan integration.
+
+Enabling Infinispan second level cache provider inside a JPA/Hibernate application that runs in single node is very straightforward.
+First, make sure the Hibernate Infinispan cache provider is available in the classpath.
+Then, modify the persistence.xml to include these properties:
+
+[source, XML, indent=0]
+----
+<!-- Use Infinispan second level cache provider -->
+<property name="hibernate.cache.region.factory_class" value="infinispan"/>
+
+<!--
+   Force using local configuration when only using a single node.
+   Otherwise a clustered configuration is loaded.
+-->
+<property name="hibernate.cache.infinispan.cfg"
```

Comment:
Yeah, it could be easy to make a mistake with the config path. Did you have anything in mind? Shorter path or a flag to set it to local hiding the config path?

Reply:
Yep, `hibernate.cache.infinispan.clustered = false` sounds reasonable.